src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs

```diff
@@ -119,77 +119,47 @@ private Task<Socket> AcceptAsyncApm(Socket acceptSocket)
 
         internal Task ConnectAsync(EndPoint remoteEP)
         {
-            var tcs = new TaskCompletionSource<bool>(this);
-            BeginConnect(remoteEP, iar =>
+            // Use ValueTaskReceive so the AwaitableSocketAsyncEventArgs can be re-used later.
+            AwaitableSocketAsyncEventArgs saea = LazyInitializer.EnsureInitialized(ref EventArgs.ValueTaskReceive, () => new AwaitableSocketAsyncEventArgs());
+
+            // We don't expect concurrent users while calling ConnectAsync.
+            if (!saea.Reserve())
```

Comment:
The other methods are taking the Apm way if they can't reserve this. I wonder why they are not allocating an AwaitableSocketAsyncEventArgs?

Reply:
There's a top-level comment:

> We also currently fall back to APM implementations
    // when the single cached instance for each of send/receive is otherwise in use; we could
    // potentially also employ a global pool from which to pull in such situations.

And also a comment:

> APM operation is less expensive than creating a new SAEA and using it only once

so, the APM is cheaper.

For Connect the !Reserve() case is very unlikely, so I don't think this optimization is needed, and we can get rid of the APM code.